nexus/db-queries/src/db/datastore/target_release.rs

```diff
@@ -0,0 +1,265 @@
+use crate::db::model::{SemverVersion, TargetRelease, TargetReleaseSource};
+use crate::db::schema::target_release;
+use crate::db::schema::target_release::dsl;
+use async_bb8_diesel::AsyncRunQueryDsl as _;
+use diesel::insert_into;
+use diesel::prelude::*;
+use nexus_types::external_api::views;
+use omicron_common::api::external::{CreateResult, Error, LookupResult};
+
+impl DataStore {
+    /// Fetch the current target release, i.e., the row with the largest
+    /// generation number.
+    pub async fn target_release_get_current(
+        &self,
+        opctx: &OpContext,
+    ) -> LookupResult<TargetRelease> {
+        opctx
+            .authorize(authz::Action::Read, &authz::TARGET_RELEASE_CONFIG)
+            .await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        // Fetch the row in the `target_release` table with the largest
+        // generation number. The subquery accesses the same table, so we
+        // have to make an alias to not confuse diesel.
+        let target_release_2 =
+            diesel::alias!(target_release as target_release_2);
+        dsl::target_release
+            .filter(dsl::generation.nullable().eq_any(target_release_2.select(
+                diesel::dsl::max(target_release_2.field(dsl::generation)),
+            )))
+            .select(TargetRelease::as_select())
+            .first_async(&*conn)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Insert a new target release row and return it. It will only become
+    /// the current target release if its generation is larger than any
+    /// existing row.
+    pub async fn target_release_insert(
+        &self,
+        opctx: &OpContext,
+        target_release: TargetRelease,
+    ) -> CreateResult<TargetRelease> {
+        opctx
+            .authorize(authz::Action::Modify, &authz::TARGET_RELEASE_CONFIG)
+            .await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+        self.transaction_retry_wrapper("set_target_release")
+            .transaction(&conn, |conn| {
+                let target_release = target_release.clone();
+                async move {
+                    insert_into(dsl::target_release)
+                        .values(target_release)
+                        .returning(TargetRelease::as_returning())
+                        .get_result_async(&conn)
+                        .await
+                }
+            })
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Convert a model-level target release to an external view.
+    /// This method lives here because we have to look up the version
+    /// corresponding to the TUF repo.
+    pub async fn target_release_view(
+        &self,
+        opctx: &OpContext,
+        target_release: &TargetRelease,
+    ) -> LookupResult<views::TargetRelease> {
+        opctx
```

Comment:
I don't think this makes any difference but I figured I'd mention it: I think in other places (like `InvOmicronZone` I mentioned elsewhere), we've used a pattern where:

- you'd have a `TargetRelease::into_external(self, system_version: SystemVersion) -> Result<views::TargetRelease, ?>`
- this function would fetch the system version like it does, but then use `target_release.into_external(system_version)`

Like I said, I'm not sure it makes any difference.  I kind of like that the conversion to/from the external type could both live with the db type rather than in this logic here but 🤷

Reply:
Thanks, refactored in 27b7e8da8.